### ISSUE TITLE
[Enhancement] Nested read-write lock issue in ManagedCursor

### ISSUE THREAD

**Issue by Denovo1998** [apache/pulsar#23605]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

The following are all related to the `org.apache.bookkeeper.mledger.impl.ManagedCursorImpl#isMessageDeleted`:
https://github.com/apache/pulsar/blob/89ccb7361b80dfb7897ea2555dfb6e884f62b916/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L3542-L3551

1. A write lock has already been acquired, but a read lock has been acquired as well. This doesn't pose a major problem, but it does result in some unnecessary additional overhead.
https://github.com/apache/pulsar/blob/89ccb7361b80dfb7897ea2555dfb6e884f62b916/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L2328-L2348

2. The performance impact of nested read locks can be negligible, but frequent acquisition and release of locks will slightly increase system overhead.
https://github.com/apache/pulsar/blob/89ccb7361b80dfb7897ea2555dfb6e884f62b916/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L1570-L1575

### Solution

**Add a method that does not require a read lock.**
```java
    private boolean isMessageDeletedWithOutReadLock(Position position) {
        return position.compareTo(markDeletePosition) <= 0
                || individualDeletedMessages.contains(position.getLedgerId(), position.getEntryId());
    }
```

### Anything else?

**I think the modification at [22246](https://github.com/apache/pulsar/pull/22246) here is for code reuse, but the nested read-write locks in these two segments may need to be modified.**

**If such a modification is necessary, please let me know, and I will submit a simple PR.**

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

**Comment by Denovo1998** [comment]
@dao-jun @lhotari PTAL.

**Comment by dao-jun** [comment]
Personally, I don't think we need to `fix` it, Read/Write Lock is ReentrantLock, and it just compare the lock `owner` of the AQS and the current thread when we required a ReadLock after required a WriteLock.

**Comment by lhotari** [comment]
> 2\. The performance impact of nested read locks can be negligible, but frequent acquisition and release of locks will slightly increase system overhead.

Makes sense. I have a large changeset to address multiple performance issues with Broker cache and one of the changes is related: https://github.com/lhotari/pulsar/commit/a0d0b544b1beb1dddb2673898969bcc2839e96b6 . I haven't yet split this large changeset into PRs for review.

**Comment by Denovo1998** [comment]
@lhotari Oh, I saw it! If you think it‘s necessary. Maybe you can solve the read-write lock nesting problem mentioned in this issue in your pr.

**Comment by lhotari** [comment]
> @lhotari Oh, I saw it! If you think it‘s necessary. Maybe you can solve the read-write lock nesting problem mentioned in this issue in your pr.

@Denovo1998 you can go ahead and create a PR if you wish.

**Comment by lhotari** [comment]
Addressed by #23609
